zproject/backends.py

```diff
@@ -164,6 +168,60 @@ def common_get_active_user(email: str, realm: Realm,
 
     return user_profile
 
+AuthFuncT = TypeVar('AuthFuncT', bound=Callable[..., Optional[UserProfile]])
+rate_limiting_rules = settings.RATE_LIMITING_RULES['authenticate_by_username']
+
+class RateLimitedAuthenticationByUsername(RateLimitedObject):
+    def __init__(self, username: str) -> None:
+        self.username = username
+
+    def __str__(self) -> str:
+        return "Username: {}".format(self.username)
+
+    def key_fragment(self) -> str:
+        return "{}:{}".format(type(self), self.username)
```

Comment:
For handling unit testing, it might be better to migrate our redis code to use a KEY_PREFIX that we can randomize with each new test class similar to memcached.